x/mongo/driver/topology/pool.go

```diff
@@ -601,3 +517,392 @@ func (p *pool) clear(serviceID *primitive.ObjectID) {
+	wait := func() (*wantConn, *connection, bool) {
+		p.connsCond.L.Lock()
+		defer p.connsCond.L.Unlock()
+
+		for !condition() {
+			p.connsCond.Wait()
+		}
+
+		if ctx.Err() != nil {
+			return nil, nil, false
+		}
+
+		p.newConnWait.cleanFront()
+		w := p.newConnWait.popFront()
+		if w == nil {
+			return nil, nil, false
+		}
+
+		conn, err := newConnection(p.address, p.opts...)
+		if err != nil {
+			w.tryDeliver(nil, err)
+			return nil, nil, false
+		}
+
+		conn.pool = p
+		conn.poolID = atomic.AddUint64(&p.nextID, 1)
+		p.conns[conn.poolID] = conn
+
+		return w, conn, true
+	}
+
+	for ctx.Err() == nil {
+		w, conn, ok := wait()
+		if !ok {
+			continue
+		}
+
+		if p.monitor != nil {
+			p.monitor.Event(&event.PoolEvent{
+				Type:         event.ConnectionCreated,
+				Address:      p.address.String(),
+				ConnectionID: conn.poolID,
+			})
+		}
+
+		conn.connect(context.Background())
+		err := conn.wait()
+		if err != nil {
+			_ = p.removeConnection(conn, event.ReasonConnectionErrored)
+			_ = p.closeConnection(conn)
+			w.tryDeliver(nil, err)
+			continue
+		}
+
+		if p.monitor != nil {
+			p.monitor.Event(&event.PoolEvent{
+				Type:         event.ConnectionReady,
+				Address:      p.address.String(),
+				ConnectionID: conn.poolID,
+			})
+		}
+
+		if w.tryDeliver(conn, nil) {
+			continue
+		}
+
+		_ = p.checkInNoEvent(conn)
+	}
+}
+
+func (p *pool) maintain(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	ticker := time.NewTicker(p.maintainInterval)
+	defer ticker.Stop()
+
+	// remove removes the *wantConn at index i from the slice and returns the new slice. The order
+	// of the slice is not maintained.
+	remove := func(arr []*wantConn, i int) []*wantConn {
+		end := len(arr) - 1
+		arr[i], arr[end] = arr[end], arr[i]
+		return arr[:end]
+	}
+
+	// removeNotWaiting removes any wantConns that are no longer waiting from given slice of
+	// wantConns. That allows maintain() to use the size of its wantConns slice as an indication of
+	// how many new connection requests are outstanding and subtract that from the number of
+	// connections to ask for when maintaining minPoolSize.
+	removeNotWaiting := func(arr []*wantConn) []*wantConn {
+		for i := len(arr) - 1; i >= 0; i-- {
+			w := arr[i]
+			if !w.waiting() {
+				arr = remove(arr, i)
+			}
+		}
+
+		return arr
+	}
+
+	wantConns := make([]*wantConn, 0, p.minSize)
+	defer func() {
+		for _, w := range wantConns {
+			w.tryDeliver(nil, ErrPoolDisconnected)
+		}
+	}()
+
+	for {
+		p.removePerishedConns()
+
+		// Remove any wantConns that are no longer waiting.
+		wantConns = removeNotWaiting(wantConns)
+
+		// Figure out how many more wantConns we need to satisfy minPoolSize. Assume that the
+		// outstanding wantConns (i.e. the ones that weren't removed from the slice) will all return
+		// connections when they're ready, so only add wantConns to make up the difference.
+		total := p.totalConnectionCount()
+		n := int(p.minSize) - total - len(wantConns)
```

Comment:
It's not safe to assume that every `wantConn` in the `newConnWait` queue will result in a new connection. In fact, most `wantConn` requests in `newConnWait` will probably be cancelled before a new connection is created because an idle connection is checked in. In that case, using the length of `newConnWait` to subtract from the number of connections requested by `maintain` would likely lead to fewer than `minPoolSize` connections in the pool.

The only `wantConn` requests we're sure won't be cancelled are those created by `maintain`, so I used that length as the "outstanding new connection requests we're sure won't be cancelled". However, as your example suggests, that can (and likely will) lead to more than `minPoolSize` connections in the pool. A simple mitigation could be to only request up to a max number of new connections at a time (e.g. 5 or 10), which would limit the possible overshoot. Does that sound like a reasonable improvement?

Reading the code again, I actually think the most correct change is to incorporate `minPoolSize` into the `createConnections` condition, allowing connections to be created with no `wantConn` until `minPoolSize` is satisfied. I will prototype that change and see if it's doable now or needs to be completed later with [GODRIVER-2131](https://jira.mongodb.org/browse/GODRIVER-2131).

Additionally, I think it's worth testing interaction between `minPoolSize` and `maxPoolSize` that could cause unexpected pool behavior. I will add tests where `minPoolSize == maxPoolSize` and `minPoolSize > maxPoolSize`.

Reply:
IIUC, we enqueue more `wantConn`s than we need, but `createConnections` won't actually end up creating more than `maxPoolSize` connections due to the `poolHasSpace` check in its `condition` function. Is this correct?

Reply:
> It's not safe to assume that every wantConn in the newConnWait queue will result in a new connection. In fact, most wantConn requests in newConnWait will probably be cancelled before a new connection is created because an idle connection is checked in. In that case, using the length of newConnWait to subtract from the number of connections requested by maintain would likely lead to fewer than minPoolSize connections in the pool.

That is a good point.

> I've updated maintain() to request a max of 10 new connections at a time to satisfy minPoolSize.

Requesting a max of 10 new connections at a time seems like a good compromise until GODRIVER-2131 is addressed.